index.js

```diff
@@ -1,135 +1,170 @@
+ * @returns {mcl.G2} g2
+ */
+function hashToG2 (messageHash, domain) {
+	const xReal = keccak256(Buffer.concat([
+    messageHash,
+    Buffer.alloc(1, domain),
+    Buffer.from([1]),
+  ]))
+  const xRealFp = new mcl.Fp()
+  xRealFp.setLittleEndian(xReal)
+	const xImag = keccak256(Buffer.concat([
+    messageHash,
+    Buffer.alloc(1, domain),
+    Buffer.from([1]),
+  ]))
+  const xImagFp = new mcl.Fp()
+  xImagFp.setLittleEndian(xImag)
+  const xCoordinate = new mcl.Fp2()
+  xCoordinate.set_a(xRealFp)
+  xCoordinate.set_b(xImagFp)
+  return xCoordinate.mapToG2()
 }
 
-// msg:string
-// domain:int
-const hash_to_G2 = (msg, domain) => {
-	let x_re = keccak256(`${msg}${domain}01`)
-	//console.log(x_re.toString('hex'))
-	let x_im = keccak256(`${msg}${domain}02`)
-	//console.log(x_im.toString('hex'))
-	//let x_coordinate = [x_re, x_im]
-
-	let done = false
-	let y_re
-	let y_im 
-
-	while(!done) {
-		let x_re_num = parseInt(x_re.toString('hex'), 16)
-		let x_im_num = parseInt(x_im.toString('hex'), 16)
-		let y_re_sq = x_re_num ** 3 + 4
-		let y_im_sq = x_im_num ** 3 + 4
-		y_re = modular_squareroot(y_re_sq)
-		y_im = modular_squareroot(y_im_sq)
-
-		if (y_re !== null && y_im !== null) done = true
-
-		done = true
-	}
-
-	return [y_re, y_im]
+/**
+ * @param {Array<bytes48>} pubkeys
+ * @returns {bytes48}
+ */
+function aggregatePubkeys(pubkeys) {
+  return Buffer.from(
+    pubkeys.map((pub) => fromBuffer(mcl.G1, pub))
+    .reduce((acc, val) => mcl.add(acc, val))
+    .serialize()
+  )
 }
 
-// s:mcl.Fr
-// msg:string
-// return sig:mcl.G2
-const sign = (s, msg, domain) => {
-	let str = `${msg}${domain}`
-	let hash = mcl.hashAndMapToG2(str)
-	return mcl.mul(hash, s)
+/**
+ * @param {Array<bytes96>} signatures
+ * @returns {bytes96}
+ */
+function aggregateSignatures(signatures) {
+  return Buffer.from(
+    signatures.map((sig) => fromBuffer(mcl.G2, sig))
+    .reduce((acc, val) => mcl.add(acc, val))
+    .serialize()
+  )
 }
 
-// list can be either pubkeys:mcl.G1[] or signatures:mcl.G2[]
-// if pubkeys, returns aggregated pubkey P:mcl.G1
-// if signatures, returns aggregated signature Sig:mc;.G2
-const aggregate = (list) => {
-	let res = list[0]
-	for (let i = 1; i < list.length; i++) {
-		res = mcl.add(res, list[i])
-	}
-	return res
+/**
+ * @param {bytes48} pubkey
+ * @param {bytes32} messageHash
+ * @param {bytes96} signature
+ * @param {int} domain
+ * @returns {boolean}
+ */
+function verify (pubkey, messageHash, signature, domain) {
+  const pubkeyG1 = fromBuffer(mcl.G1, pubkey)
+  const signatureG2 = fromBuffer(mcl.G2, signature)
+	return mcl.pairing(pubkeyG1, hashToG2(messageHash, domain)).isEqual(mcl.pairing(g1(), signatureG2))
 }
 
-// pubkey:bls.PublicKey()
-// message:string
-// signature:bls.Signature()
-// domain:uint
-const verify = (pub, message, sig, domain) => {
-	if (!(pub instanceof mcl.G1)) throw new Error('verify:wrong pubkey type')
-	if (!(sig instanceof mcl.G2)) throw new Error('verify:wrong signature type')
```

Comment:
should we keep these checks?

Reply:
Changed the verify function to take Buffers as input.
Then we create the G1 and G2 objects inside the verify function.

Perhaps we can do a length check on the input Buffers? make sure they are 48 and 96 bytes long?

Reply:
yeah, that would be good in case of invalid input.